File: py_src/taskito/dashboard/url_safety.py
"""Outbound URL safety checks for dashboard-configured webhooks.

We refuse to deliver to loopback, link-local, and RFC1918 addresses by
default — an operator who can write to ``dashboard_settings`` could
otherwise turn the worker into an SSRF proxy. The ``TASKITO_WEBHOOKS_ALLOW_PRIVATE``
environment variable disables the guard for local development.
"""

from __future__ import annotations

import ipaddress
import os
import socket
import urllib.parse

# Hostnames that always resolve to loopback / never-leave-this-host regardless
# of DNS, but might be missed by a strict ``ipaddress.is_private`` check.
_BLOCKED_HOSTNAME_SUFFIXES = (
    ".localhost",
    ".local",
    ".internal",
    ".intranet",
    ".lan",
    ".private",
)
_BLOCKED_HOSTNAMES = frozenset(
    {"localhost", "localhost.localdomain", "ip6-localhost", "ip6-loopback"}
)

_ALLOW_ENV_VAR = "TASKITO_WEBHOOKS_ALLOW_PRIVATE"


class UnsafeWebhookUrl(ValueError):
    """Raised when a webhook URL targets an address we won't deliver to."""


def _is_private_ip(ip: str) -> bool:
    try:
        address = ipaddress.ip_address(ip)
    except ValueError:
        return False
    return (
        address.is_private
        or address.is_loopback
        or address.is_link_local
        or address.is_multicast
        or address.is_reserved
        or address.is_unspecified
    )


def _hostname_is_blocked(hostname: str) -> bool:
    lowered = hostname.lower()
    if lowered in _BLOCKED_HOSTNAMES:
        return True
    return any(lowered.endswith(suffix) for suffix in _BLOCKED_HOSTNAME_SUFFIXES)


def validate_webhook_url(url: str) -> None:
    """Reject ``url`` if it targets a private/loopback/link-local destination.

    Set ``TASKITO_WEBHOOKS_ALLOW_PRIVATE=1`` in the environment to disable
    the guard (intended for local development against ``http://localhost``).

    Raises:
        UnsafeWebhookUrl: on scheme other than http/https, missing host, or
            a host that resolves to a private/loopback IP.
    """
    parsed = urllib.parse.urlparse(url)
    if parsed.scheme not in ("http", "https"):
        raise UnsafeWebhookUrl(f"URL scheme must be http or https, got {parsed.scheme!r}")
    if not parsed.hostname:
        raise UnsafeWebhookUrl("URL must include a hostname")

    if os.environ.get(_ALLOW_ENV_VAR):
        return

    hostname = parsed.hostname
    if _hostname_is_blocked(hostname):
        raise UnsafeWebhookUrl(f"URL host {hostname!r} resolves to a private network")

    # Literal IPs are checked directly; named hosts are resolved.
    try:
        ipaddress.ip_address(hostname)
        addresses: list[str] = [hostname]
    except ValueError:
        try:
            addresses = [
                str(info[4][0])
                for info in socket.getaddrinfo(hostname, None, type=socket.SOCK_STREAM)
            ]
        except OSError as e:
            raise UnsafeWebhookUrl(f"could not resolve {hostname!r}: {e}") from None

    for ip in addresses:
        if _is_private_ip(ip):
            raise UnsafeWebhookUrl(f"URL host {hostname!r} resolves to private address {ip}")


def is_safe_redirect(path: str | None) -> bool:
    """Whether ``path`` is safe to use as a post-login same-origin redirect target.

    Accepts only relative paths rooted at ``/``. Rejects absolute URLs
    (``http://evil.com/x``), protocol-relative URLs (``//evil.com/x``),
    and anything without a leading slash. Empty / ``None`` is rejected so
    the caller can fall back to a default explicitly.
    """
    if not path:
        return False
    if not path.startswith("/"):
        return False
    if path.startswith("//") or path.startswith("/\\"):
        return False
    parsed = urllib.parse.urlparse(path)
    return not (parsed.scheme or parsed.netloc)

File: sdks/node/src/webhooks/urlSafety.ts
import { lookup } from "node:dns/promises";
import { isIP } from "node:net";
import { UnsafeWebhookUrlError } from "./errors";

/**
 * Outbound URL safety checks for dashboard-configured webhooks (SSRF guard).
 *
 * Anyone who can write a webhook subscription could otherwise point it at
 * `http://169.254.169.254/...` and turn the worker into an SSRF proxy for cloud
 * metadata, loopback services, or the RFC1918 intranet. Delivery to loopback,
 * link-local, RFC1918, CGNAT, and ULA destinations is refused by default; set
 * `TASKITO_WEBHOOKS_ALLOW_PRIVATE` (truthy) to disable the guard when developing
 * against a local endpoint.
 *
 * Mirrors the cross-SDK `validate_webhook_url` contract.
 */

const ALLOW_ENV_VAR = "TASKITO_WEBHOOKS_ALLOW_PRIVATE";

// Names that never leave this host regardless of DNS, which an address check
// (run only after resolution) would miss for names that fail to resolve.
const BLOCKED_HOSTNAMES = new Set([
  "localhost",
  "localhost.localdomain",
  "ip6-localhost",
  "ip6-loopback",
]);
const BLOCKED_SUFFIXES = [".localhost", ".local", ".internal", ".intranet", ".lan", ".private"];

const FALSY_ENV_VALUES = new Set(["", "0", "false", "no", "off"]);

/** Destinations that never belong to a third-party HTTP endpoint. */
const BLOCKED_V4_CIDRS: ReadonlyArray<readonly [string, number]> = [
  ["0.0.0.0", 8], // this-network / unspecified
  ["10.0.0.0", 8],
  ["100.64.0.0", 10], // CGNAT
  ["127.0.0.0", 8], // loopback
  ["169.254.0.0", 16], // link-local, incl. cloud metadata
  ["172.16.0.0", 12],
  ["192.0.0.0", 24], // IETF protocol assignments
  ["192.0.2.0", 24], // documentation
  ["192.168.0.0", 16],
  ["198.18.0.0", 15], // benchmarking
  ["198.51.100.0", 24], // documentation
  ["203.0.113.0", 24], // documentation
  ["224.0.0.0", 4], // multicast
  ["240.0.0.0", 4], // reserved + broadcast
];

const BLOCKED_V6_CIDRS: ReadonlyArray<readonly [string, number]> = [
  ["::", 128], // unspecified
  ["::1", 128], // loopback
  ["fc00::", 7], // unique-local
  ["fe80::", 10], // link-local
  ["ff00::", 8], // multicast
  ["2001:db8::", 32], // documentation
];

export interface UrlSafetyOptions {
  /** Bypass the private-address checks. Defaults to the environment variable. */
  allowPrivate?: boolean;
}

/**
 * Validate `raw` without resolving DNS.
 *
 * Used on the registration path, which is synchronous: it rejects a bad scheme,
 * a missing host, a known-local name, and a literal private address right away.
 * Named hosts are fully checked by {@link assertSafeWebhookUrl} at delivery time.
 */
export function assertSafeWebhookUrlSync(raw: string, options?: UrlSafetyOptions): void {
  const hostname = parseHostname(raw);
  if (allowPrivate(options)) {
    return;
  }
  assertHostnameAllowed(hostname);
  if (isIP(hostname) !== 0) {
    assertAddressAllowed(hostname, hostname);
  }
}

/**
 * Validate `raw`, resolving the host and checking every address it maps to.
 *
 * Called on every delivery attempt, not just at registration: a name that was
 * safe when the webhook was created can be rebound to a private address later.
 * A residual race remains between this resolve and the socket connect, but it
 * closes the wide registration-to-delivery window.
 *
 * @throws UnsafeWebhookUrlError on a non-http(s) scheme, a missing host, an
 *   unresolvable host, or a host that maps to a blocked address.
 */
export async function assertSafeWebhookUrl(raw: string, options?: UrlSafetyOptions): Promise<void> {
  const hostname = parseHostname(raw);
  if (allowPrivate(options)) {
    return;
  }
  assertHostnameAllowed(hostname);

  if (isIP(hostname) !== 0) {
    assertAddressAllowed(hostname, hostname);
    return;
  }
  let resolved: Array<{ address: string }>;
  try {
    resolved = await lookup(hostname, { all: true, verbatim: true });
  } catch {
    throw new UnsafeWebhookUrlError(`could not resolve webhook host ${hostname}`);
  }
  for (const { address } of resolved) {
    assertAddressAllowed(address, hostname);
  }
}

/** Whether the guard is disabled, from the option or the environment. */
function allowPrivate(options?: UrlSafetyOptions): boolean {
  if (options?.allowPrivate !== undefined) {
    return options.allowPrivate;
  }
  const value = process.env[ALLOW_ENV_VAR];
  return value !== undefined && !FALSY_ENV_VALUES.has(value.trim().toLowerCase());
}

/**
 * Parse `raw` and return its hostname (IPv6 brackets stripped).
 *
 * Scheme and host are validated even when the guard is bypassed, so a
 * misconfigured URL gets the same feedback either way.
 */
function parseHostname(raw: string): string {
  let url: URL;
  try {
    url = new URL(raw);
  } catch {
    throw new UnsafeWebhookUrlError("webhook url is not a valid URL");
  }
  if (url.protocol !== "http:" && url.protocol !== "https:") {
    throw new UnsafeWebhookUrlError("webhook url scheme must be http or https");
  }
  if (!url.hostname) {
    throw new UnsafeWebhookUrlError("webhook url must include a host");
  }
  return stripBrackets(url.hostname);
}

/** `URL.hostname` keeps the brackets around an IPv6 literal; drop them. */
function stripBrackets(hostname: string): string {
  return hostname.startsWith("[") && hostname.endsWith("]") ? hostname.slice(1, -1) : hostname;
}

function assertHostnameAllowed(hostname: string): void {
  const lowered = hostname.toLowerCase();
  const blocked =
    BLOCKED_HOSTNAMES.has(lowered) || BLOCKED_SUFFIXES.some((suffix) => lowered.endsWith(suffix));
  if (blocked) {
    throw new UnsafeWebhookUrlError(`webhook url host ${hostname} is a local-only name`);
  }
}

function assertAddressAllowed(address: string, hostname: string): void {
  if (isBlockedAddress(address)) {
    throw new UnsafeWebhookUrlError(
      `webhook url host ${hostname} resolves to the private or reserved address ${address}`,
    );
  }
}

function isBlockedAddress(address: string): boolean {
  const bytes = ipBytes(address);
  if (!bytes) {
    // An address we cannot parse is not an address we are willing to POST to.
    return true;
  }
  if (bytes.length === 4) {
    return matchesAny(bytes, BLOCKED_V4_CIDRS);
  }
  // IPv4-mapped IPv6 (::ffff:a.b.c.d) must be judged as the v4 address it carries.
  const mapped = mappedV4(bytes);
  return mapped ? matchesAny(mapped, BLOCKED_V4_CIDRS) : matchesAny(bytes, BLOCKED_V6_CIDRS);
}

function matchesAny(bytes: Uint8Array, cidrs: ReadonlyArray<readonly [string, number]>): boolean {
  return cidrs.some(([network, prefixLength]) => {
    const networkBytes = ipBytes(network);
    return networkBytes ? inNetwork(bytes, networkBytes, prefixLength) : false;
  });
}

/** Whether `bytes` shares its first `prefixLength` bits with `network`. */
function inNetwork(bytes: Uint8Array, network: Uint8Array, prefixLength: number): boolean {
  if (bytes.length !== network.length) {
    return false;
  }
  const wholeBytes = prefixLength >> 3;
  for (let index = 0; index < wholeBytes; index += 1) {
    if (bytes[index] !== network[index]) {
      return false;
    }
  }
  const remainingBits = prefixLength & 7;
  if (remainingBits === 0) {
    return true;
  }
  const mask = 0xff << (8 - remainingBits);
  return ((bytes[wholeBytes] ?? 0) & mask) === ((network[wholeBytes] ?? 0) & mask);
}

/** The embedded IPv4 address of an `::ffff:a.b.c.d` mapping, if this is one. */
function mappedV4(bytes: Uint8Array): Uint8Array | undefined {
  const prefix = Uint8Array.from([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0xff, 0xff]);
  for (let index = 0; index < prefix.length; index += 1) {
    if (bytes[index] !== prefix[index]) {
      return undefined;
    }
  }
  return bytes.slice(12);
}

/** Raw bytes of an IPv4/IPv6 literal; `undefined` when `ip` is not an address. */
function ipBytes(ip: string): Uint8Array | undefined {
  const family = isIP(ip);
  if (family === 4) {
    return Uint8Array.from(ip.split(".").map(Number));
  }
  return family === 6 ? v6Bytes(ip) : undefined;
}

/** Expand a (possibly `::`-compressed) IPv6 literal to its 16 bytes. */
function v6Bytes(ip: string): Uint8Array | undefined {
  const halves = ip.split("::");
  if (halves.length > 2) {
    return undefined;
  }
  const head = v6Groups(halves[0] ?? "");
  const tail = halves.length === 2 ? v6Groups(halves[1] ?? "") : [];
  if (!head || !tail) {
    return undefined;
  }
  const gap = 16 - head.length - tail.length;
  if (halves.length === 1 ? gap !== 0 : gap < 0) {
    return undefined;
  }
  return Uint8Array.from([...head, ...new Array<number>(gap).fill(0), ...tail]);
}

function v6Groups(part: string): number[] | undefined {
  if (!part) {
    return [];
  }
  const bytes: number[] = [];
  for (const group of part.split(":")) {
    // A trailing dotted quad (::ffff:1.2.3.4) contributes four bytes.
    if (group.includes(".")) {
      if (isIP(group) !== 4) {
        return undefined;
      }
      bytes.push(...group.split(".").map(Number));
      continue;
    }
    const value = Number.parseInt(group, 16);
    if (!Number.isInteger(value) || value < 0 || value > 0xffff) {
      return undefined;
    }
    bytes.push(value >> 8, value & 0xff);
  }
  return bytes;
}
